### PR TITLE
Signup A/B Test: Use progress bar instead of # steps

### PR DESCRIPTION
**Pull request by aidvu**
Use a visual progress bar, instead of the step numbers.

It's both a better UX and users won't get confused when we skip steps so they go from 2 to 4.

Test existing flows. Make sure the progress bar makes sense and nothing is broken.

Before:
steps

After:
progress bar

### CODE REVIEW

**Comment by folletto** [comment]
> It's both a better UX and users won't get confused when we skip steps so they go from 2 to 4.

I'm not sure it's better for UX. 

The semantics of the progress bar are very different from the semantics of a wizard progress. One is automated, the other is user driven. One is meant to indicate movement and work, the other just next step.

For example:

screen shot 2017-05-25 at 10 02 11

This empty line is hard to read. What does it mean?

Then, this:

cap-drop

Which is animated (as a progress bar should, signifying it's _working_), and distracts from the page.

---

If we think the steps indication is confusing, then we should design a control ad-hoc for step processes like this one.

But first: what was the rationale behind this suggestion?

**Comment by bisko** [comment]
👍  for @folletto's comment about the animated progress - this is not something that's currently working. It should be solid color and with less contrast if possible. It should be a visual cue, not something the user will interact with.

Also 👍  for the empty progress :) A much better option would be to start at 1, instead of 0 progress. This will show that the user is at least somewhere. This would mean that the total progress would be  ( `steps` + 1 ), instead of `steps`.

@folletto a bit of background on the rationale ( @aidvu correct me if I'm wrong ). The change to a progress bar instead of a numbered step counter was discussed in the past, but it didn't get into focus to get implemented.

The Signup process consists of a number of steps, which is let's say 6. If the user is going through the flow they'll see: `Step 2 of 6`. 
The problem comes when we want the user to skip steps, e.g. optional steps, which @aidvu mentions above. This would make the user see a jump from `Step 2 of 6` to `Step 4 of 6` without any indication why the jump happened. Is there a fork in the logic?

In the past we also implemented some AB tests that would dynamically insert steps in the signup flow while the user was progressing through the flow ( not at the start ), which would give the user a jump from `Step 2 of 6` to `Step 3 of 7`. It looks confusing. 

The progress bar will still have a similar issue, but it's more "masked" in similar changes in the flow. For example jumping from step 2 to step 3, and changing the total step amount from 6 to 7, will still increment the progress bar, although with a smaller amount. This still feels like progress, instead of seeing that you have to complete yet another step.

**Comment by folletto** [comment]
> It looks confusing.

Did any user ever complained about it? 
I'm asking to understand if this is a user problem or something we want to clean up ourselves (both valid, difference in priorities and how to fix).

> This still feels like progress, instead of seeing that you have to complete yet another step.

Ok, makes sense. So we'd like to have some kind of visual indication if progress that however doesn't "lock" in discrete steps, is that right?

Was discussed about... just removing it?

**Comment by aidvu** [comment]
Issue here: https://github.com/Automattic/wp-calypso/issues/13449. 

Mostly reported internally. It's a bit subtle, so I wouldn't expect users to report such things as they are (in my opinion) masked by the feeling you get for creating your site. :) 

> Ok, makes sense. So we'd like to have some kind of visual indication if progress that however doesn't "lock" in discrete steps, is that right?
Yep!

As for removal... It's the easiest way, but I think knowing how much you have left might give you a nudge to finish, now that you're so close to creating your website. :) I personally like to see how much I have to go through when going through any kind of signup.

**Comment by aidvu** [comment]
Did design pairing with @folletto, and we'll create a new component.

Looks: 
alt-7

**Comment by mikeshelton1503** [comment]
+1 to @folletto's comments here.

I've seen research over the years of the various affects of progress bars/indicators on conversion rates. The general consensus is they improve conversion rates, so I'd definitely vote to keep some form of progress indicator. The part that is less clear and varies widely is what style of progress indicator performs best.

Given that this change affects all steps of our signup funnel, I highly recommend that we test this change 50/50 against the current version. Also, lets keep in mind other tests going on in the NUX flow to not add additional variables to those tests cc @fditrapani 

>Did design pairing with @folletto, and we'll create a new component.

I think the new design improves on some aspects and is on the right track. It feels a bit small though, almost looking like a decorative rule instead of a progress bar.

I'm happy to take a stab here too.

**Comment by folletto** [comment]
> I think the new design improves on some aspects and is on the right track. It feels a bit small though, almost looking like a decorative rule instead of a progress bar.

Yep, I was aiming for that: I didn't want to shift the actual focus of the page on the forms.

> I'm happy to take a stab here too.

Absolutely, we can iterate up. 

---

Meanwhile, I just pushed a few tweaks, so just in case the current proposal is ready if we want to ship:

screen shot 2017-05-25 at 14 37 36

**Comment by folletto** [comment]
Quick alternative, just for brainstorming, an arrow:

screen shot 2017-05-25 at 14 43 59

**Comment by aidvu** [comment]
Ok, talked with @lucasartoni, plan to test it `2017-06-05`. Will update the PR.

**Comment by drw158** [comment]
It's getting better with each iteration! The bar at 1px(?) feels too delicate. 2px would be a lot better I think. The arrow is much better than a diamond, which feels out of place. I'd probably stick with a circle since it fits with our other UI elements. Or maybe use a gridicon arrow.

**Comment by folletto** [comment]
> I'd probably stick with a circle since it fits with our other UI elements. Or maybe use a gridicon arrow.

I tried both: the circle looks too much like a "Puck", so something the user can drag (we have a similar control on the media library) so I didn't do that.

The gridicon arrow... it felt blurry due to size, hence I crafted a custom arrow. I think we can go for that.

> The bar at 1px(?) feels too delicate. 2px would be a lot better I think.

1px felt sharper: the tip of the arrow is 1px, thus as sharp as it can. 2px would make the tip 2px, which is a bit blurry. I'd be ok either way, was mostly a preference of the moment. Give it a try and see :)

**Comment by aidvu** [comment]
This is open for updates until the A/B test launch date, so feel free to chime in with whatever you think is better. :) :bowing_man:

File: client/components/wizard-bar/docs/example.jsx [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ */
+import React from 'react';
+import PureRenderMixin from 'react-pure-render/mixin';
+
+/**
+ * Internal dependencies
+ */
+import WizardBar from 'components/wizard-bar';
+
+module.exports = React.createClass( {
```
Comment (klimeryk) [inline client/components/wizard-bar/docs/example.jsx]: BTW, I don't see this component in devdocs 😅

File: client/components/wizard-bar/progress-icon.jsx [hunk 2]
```diff
@@ -0,0 +1,10 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
+
+export default () => (
```
Comment (klimeryk) [inline client/components/wizard-bar/progress-icon.jsx]: Could you do something like:
```
const ProgressIcon = () => { }
export default ProgressIcon;
```
instead? That way `displayName` is set by babel and it's easier to debug and inpect components. I doubt this one will cause anyone any troubles, but it's just good practice - otherwise it shows up as `Unknown` in React DevTools:
screen shot 2017-05-25 at 20 20 55

File: client/components/wizard-bar/index.jsx [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+		if ( this.props.color ) {
+			styles.backgroundColor = this.props.color;
+		}
+
+		return (
+			<div>
+				<div className="wizard-bar__progress" style={ styles } />
+				<div
+					className="wizard-bar__indicator"
+					style={ { width: ( completionPercentage + 1 ) + '%' } }
```
Comment (klimeryk) [inline client/components/wizard-bar/index.jsx]: Initially I thought this would cause problems, but apparently `width: 101%` is perfectly fine and makes the design work 😀 TIL

File: client/components/wizard-bar/index.jsx [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+				>
+					{ <ProgressIcon /> }
+				</div>
+			</div>
+		);
+	}
+
+	render() {
+		const classes = classnames( this.props.className, 'wizard-bar' );
```
Comment (klimeryk) [inline client/components/wizard-bar/index.jsx]: `className` doesn't seem to be in the `propTypes` and is not used anywhere in existing code - is it a leftover from an older version of the code? I think it could be removed (and re-added if it'll be really needed).

File: client/components/wizard-bar/progress-icon.jsx (hunk 2 above)
Comment (aidvu) [inline client/components/wizard-bar/progress-icon.jsx]: Nice point! Should update those for the existing site flow.

File: client/components/wizard-bar/docs/example.jsx (hunk 1 above)
Comment (aidvu) [inline client/components/wizard-bar/docs/example.jsx]: I guess it appears when built? Or should I add it somewhere else? :)

Comment (klimeryk) [inline client/components/wizard-bar/docs/example.jsx]: Honestly, no idea - I couldn't find any docs on it in the repo :/ `WizardBar`'s `README.md` is searchable through devdocs, so it is picked up, but it's not being added to the UI Components page 🤔

**Comment by keoshi** [comment]
Probably adding a bit on entropy to the conversation at this point, but just found about this thread so bear with me.

To me the issue would be that the entire process is hard to grasp at any step, because I can't tell how many steps there are or what progress have I made. I understand the edge cases of introducing an additional step in between, but sounds like it could be an implementation tweak rather than a design overhaul (i.e.: accounting for that step when the process begins).

For me, as user, I'd rather have the initial solution — _Step 2 of 4_ — since it's the one that better sets the expectations.

Having said that, I think there's value in translating those steps into something more visual: to be transparent and create those expectations upfront. Leaving here some options, but feel free to discard them if they don't respond to the problem:

image

**Comment by ashleighaxios** [comment]
Don't love this solution, but curious to see how it does in a test. (What will you be measuring for success in the A/B test?) I don't _think_ everyone will know what this is supposed to indicate. _My hunch is that we're likely skewed by the process, seeing it iterate from something else; having a clear previous point of reference (IE: curse of knowledge)._ 

For the above, I'm also curious how we're handling the a11y needs of a screenreader. They should really get all of the same information a user with full sight does.

While I hate adding text back, here are a couple of other ideas that might be a bit more fair to a user with a disability.

steps_01
steps_02

Also, I can imagine the mark animating into the logo when it's done before moving to the next screen, but that's just a nice-to-have idea.

**Comment by aidvu** [comment]
@keoshi thanks for the feedback. The problem here is that we have flows that either skip steps, or even some that dynamically add steps. The idea behind is to hide the # of steps while still giving the users an idea of the progress.

> Don't love this solution, but curious to see how it does in a test. (What will you be measuring for success in the A/B test?)

@ashleighaxios I don't think we can measure much apart from the conversion rates.

> For the above, I'm also curious how we're handling the a11y needs of a screenreader. They should really get all of the same information a user with full sight does.

Very good point we have completely overlooked! 

As I'm not a designer, I'm here to implement whatever we decide would be the best to try and iterate on the progress indicator, or in the end decide that (after or before testing) the current steps solution we have is best.

**Comment by keoshi** [comment]
> The problem here is that we have flows that either skip steps, or even some that dynamically add steps.

My point was exactly in that context. As a user, I'd rather have an adaptive progress bar than one that offers no real sense of expectations or advancement.

For the particular flows you mention, wouldn't it be possible to have complete and separate flows, rather than just one that is modified? Meaning we'd start all flows in step 1, and not introduce or skip steps once the site creation process has started?

**Comment by aidvu** [comment]
>My point was exactly in that context. As a user, I'd rather have an adaptive progress bar than one that offers no real sense of expectations or advancement.

That's what I'd like to test, tho I agree that if possible, exact step # would be best.

>For the particular flows you mention, wouldn't it be possible to have complete and separate flows, rather than just one that is modified? Meaning we'd start all flows in step 1, and not introduce or skip steps once the site creation process has started?

Not with what we have now. Might be possible and would probably be interesting if we could also name the steps, so users know what they are skipping/is already finished.

**Comment by drw158** [comment]
I like @ashleighaxios's [solution](https://github.com/Automattic/wp-calypso/pull/14443#issuecomment-304317209) the best, if we determine that a visual progress bar is better than the plain text step counter we have now. 

screen shot 2017-05-26 at 2 37 09 pm

If we can figure this out:

> wouldn't it be possible to have complete and separate flows, rather than just one that is modified?

Then I think one of @keoshi's [designs](https://github.com/Automattic/wp-calypso/pull/14443#issuecomment-304317132) would be best (with text or something added for accessibility):

screen shot 2017-05-26 at 2 36 59 pm

**Comment by bisko** [comment]
> wouldn't it be possible to have complete and separate flows, rather than just one that is modified?

I want to add a note on why knowing the exact number of steps might be a problem. We pondered over that question a bit when we implemented #8280. 

A relevant quote from the [discussion there](https://github.com/Automattic/wp-calypso/pull/8280/files#r81229586): 

> For testing new steps, we don't want to set the variation until the last possible moment, which is right before the step (in this case - when the previous one ends). If we set the variation in the beginning it may skew the test with all the people that will drop off Signup even before reaching the step. This will require us to run the test longer, so we can gather more accurate data. If we set the variation right before the step, the only users that will get it will be users that reach the step and it will more accurately measure the step success and attrition rates.

The same is valid when jumping through steps or just not following the flow which is written in the config. If the flow isn't followed strictly and there are AB tests, their results will possibly be skewed. 

That's what we tried to account for when we worked on inserting a step into the flow ( the Site Title step in that case ), which lead us to firing the AB test and inserting the step at the last possible moment, which happens mid-flow. 

If we stick to an exact number of steps counter, this will be visible to the user and will change the amount of steps mid-flow.

That's why I think a percentage-based ([#](https://cloud.githubusercontent.com/assets/4389/26452387/b6da309e-4157-11e7-8b4b-e60aa82ee5d8.png), [#](https://cloud.githubusercontent.com/assets/618551/26509979/d77f29b6-4220-11e7-9954-8abf5437be0d.png))  progress makes more sense in this case. This way it wouldn't be immediately obvious that something is happening behind the scenes, as I mentioned above. 

Another note here - I don't think we have any dynamic step insertions or removals for the moment, so this isn't that big of a problem, but it would be awesome if we can make one design that works with both - "normal" flows, without any forks or jumps and dynamic flows that have insertion/deletion and jumps.

At some point in the future, if we manage to find a way to reliably run AB tests with dynamic number of steps, which are defined at the start, then an exact step counter would be even more awesome!

**Comment by aidvu** [comment]
I'd suggest adding a `Current Progress: #%` above the progress bar @folletto did for the test. Depending on the results, we iterate more.

**Comment by folletto** [comment]
Ok this discussion is exploding a little bit with a lot of sub threads, let's try to give some momentum back :)



## Step back in history...

We designed the current NUX Funnel in a highly modular way, hoping there would no need of branching inside. We have many flows ([config file here](https://github.com/Automattic/wp-calypso/blob/5da1ab27c5e608046377c1726547ac98ae0d80fc/client/signup/config/flows.js#L53)), and each is ideally selected to be optimized for the specific scenario. 

Here's where we meet a boundary: in practice it still means that there are some instances where this still requires some shuffling with steps.


## About testing funnels....

> For testing new steps, we don't want to set the variation until the last possible moment, which is right before the step (in this case - when the previous one ends). If we set the variation in the beginning it may skew the test with all the people that will drop off Signup even before reaching the step.

Maybe it's me, but I'm not 100% convinced on this assumption. Testing two identical funnels side-by-side with one step replaced should lead to identical results as testing one funnel with a single swap, and you can still measure accuracy on the single step (drops). That's why we designed it to be modular at funnel level.

> If we set the variation right before the step, the only users that will get it will be users that reach the step and it will more accurately measure the step success and attrition rates.

This is actually an issue imho. Funnels should be evaluated in its entirety, to make sure that there's no cumulative effect of the choices done. It's ok to measure a single step drop-off, but as we seen ourselves first hand, even just changing the order changes the drop-off rates, so it would be a _major_ problem evaluating a step in isolation.

I might be missing something, but if not, I'd suggest to reconsider the A/B tests done to funnel as new funnels, not as step-swapping inside funnels.

## What about this PR...

This PR starts specifically to find a way to visualize progress _"in the absence of a precise step measure"_. If we agree that we shouldn't drop the concept of steps, and I agree that's the _ideal scenario_, then I'd personally rather **keep the text: it's clearer, accessible, and well balanced on the page.**

Yet, even if we keep the text and find ways to avoid steps mangling, I still think this PR, now that we changed to run it as an A/B test, should be tested as is, as it would help us in validating a separate assumption that everyone here (including me) is making: steps are better. 

I'd even suggest later to test a variation without the progress indicator at all, to see how it perform.

_tl;dr: let's test the assumption of this PR, and review later with data._

**Comment by folletto** [comment]
screen shot 2017-05-29 at 11 44 48

Following the feedback above, updated SVG to be an arrow (like our `gridicon-dropdown` turned).

**Comment by fditrapani** [comment]
I noticed how the numbers in the steps get all wonky when you go back and forward some times and wanted to make a change myself. Thanks @aidvu for kicking this off. The "progress" on this thread has been great :p Nice call on adding the text @ashleighaxios.

> I'd suggest adding a Current Progress: #% above the progress bar @folletto did for the test. Depending on the results, we iterate more.

Not a huge fan of the arrow but I like this because it takes the least amount of space. One suggestion would be to change the copy so it focuses on the goal:

image

> What will you be measuring for success in the A/B test?
I suggest we look at overall conversion of the flow and also a step by step break down. @aidvu do you have a test post ready for this? I can help you get one started. 

> Plan to test it `2017-06-05`
As @mikeshelton1503 mentioned, we do have a number of tests running in the signup flow. There will be some overlap if you run it on the 5th. Can we start it on the 12th?

**Comment by aidvu** [comment]
@fditrapani I'll get the post ready before launch.

@lucasartoni we ok to move this +1 week? 

I agree with @folletto to test as is, and then iterate.

**Comment by aidvu** [comment]
Moving launch to `2017-06-12`. Will keep the test as is, then based on the results we iterate.

File: client/devdocs/design/index.jsx [hunk 5]
```diff
@@ -168,6 +169,7 @@ let DesignAssets = React.createClass( {
 					<TokenFields />
 					<VerticalMenu />
 					<Version />
+					<WizardBar />
```
Comment (klimeryk) [inline client/devdocs/design/index.jsx]: :1st_place_medal:

**Comment by fditrapani** [comment]
> And I'm not sure that's a good experience for a user that reads right to left - but that's probably a bigger issue (?) with signup overall since the Back button at the bottom should probably get flipped too, etc. Not sure if it's something that was overlooked so far, or a deliberate decision - @fditrapani, maybe you could shed some light on this? 🙏 

@klimeryk these are good points. The decisions predate me but I'd assume they were simply overlooked. We can consider alternatives after the test. As for the back button, I question whether we even need it since it duplicates browser functionality. But as you said, that's for a separate discussion PR :D

> Either way, it's not a blocker and should be addressed (if it's indeed an issue) in a separate PR.
+1

**Comment by lucasartoni** [comment]
I am very confused.
This is a PR for an A/B test, which has the purpose of validating a hypothesis.
I see a long and articulated discussion but no hypothesis, and this is a problem.

Discussing visual consistency, assumptions, LTR, etc. etc. is very premature. There are two options:

1: this is not an A/B test, then it's fine and we remove the A/B test code, we keep working on it as a new development of the interface, discussing every single aspect of it.

2: it's an A/B test: we define a clear hypothesis and the only focus is designing the fastest, cheapest, most effective test to validate/invalidate it. Once the results are in, we discuss all the consequences of the real implementation to production.

Both options at the same time are expensive and inconclusive.

**Comment by folletto** [comment]
> The decisions predate me but I'd assume they were simply overlooked.

As the designer of the original NUX Funnel, I'd say they were overlooked, or, better, we expected to work as in the current automated stack which automatically build RTL CSS: sometimes there are tweaks to be made manually, it's probably one of these cases (arrow control included).

Button icons should also swap similarly automatically, so maybe that control isn't using the button component?  Changing it might solve the problem.

But I agree even this is for a separate discussion.
